### ISSUE TITLE
Issues getting container name into HCONTAINERID

### ISSUE THREAD

**Issue by juwi** [jangaraj/Zabbix-Docker-Monitoring#17]
Hi,

I've just installed your module today. Like it, so far. However I seem to have an issue with getting the container name into HCONTAINERID. I'm on CentOS 7, so I added zabbix to the dockerroot group and changed ownership of the docker socket to root:dockerroot. As zabbix user I can run all Zabbix commands without issues. 
I Started a container with both, hostname and name set, however when I do a lookup I always get the short id as name. I also compiled the mesos branch and tried via an ENV var. Even there, I only get the ID: 
<code>
docker.discovery[Config,Env,MOD=]
ZBXD{
	"data":[
		{
			"{#FCONTAINERID}":"a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f",
			"{#HCONTAINERID}":"a8115e9632e7"},
		{
			"{#FCONTAINERID}":"039966e28e6aa67cc89355c36e50ba16a4515b70b175026eed68be7b1e26e7c5",
			"{#HCONTAINERID}":"039966e28e6a"},
		{
			"{#FCONTAINERID}":"307742a66655ade0d3bcad983e955bb32ee13987d4129c52a411083a140ad089",
			"{#HCONTAINERID}":"307742a66655"},
		{
			"{#FCONTAINERID}":"1e141d3b75d2bc09c70df0938dda74b54ac560209220b8d1cbc2369e3425cd5d",
			"{#HCONTAINERID}":"1e141d3b75d2"}]}Connection closed by foreign host.

</code>
Here is the output of inspect on the machine:
<code>
[{
    "AppArmorProfile": "",

    "Args": [
        "-c",
        "/etc/supervisord.conf"
    ],
    "Config": {
        "AttachStderr": true,
        "AttachStdin": false,
        "AttachStdout": true,
        "Cmd": [
            "/usr/bin/supervisord",
            "-c",
            "/etc/supervisord.conf"
        ],
        "CpuShares": 4096,
        "Cpuset": "",
        "Domainname": foo.local",
        "Entrypoint": null,
        "Env": [
            "INST=foobar",
            "MOD=mod-foo",
            "CONFIG=foobar",
            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
            "SERF_HOME=/usr/local/serf",
            "TERM=xterm"
        ],
        "ExposedPorts": {
        },
        "Hostname": "mod-foo",
        "Image": "registry/foo:latest",
        "Labels": {},
        "MacAddress": "",
        "Memory": 2147483648,
        "MemorySwap": 0,
        "NetworkDisabled": false,
        "OnBuild": null,
        "OpenStdin": false,
        "PortSpecs": null,
        "StdinOnce": false,
        "Tty": false,
        "User": "",
        "Volumes": null,
        "WorkingDir": ""
    },
    "Created": "2015-07-28T14:14:56.392591344Z",
    "Driver": "devicemapper",
    "ExecDriver": "native-0.2",
    "ExecIDs": [
        "60bab9a3174e2b1092c4ebad9c05a112a84a12ac76c385a4374612bbbd9d6af0"
    ],
    "HostConfig": {
        "Binds": null,
        "CapAdd": null,
        "CapDrop": null,
        "CgroupParent": "",
        "ContainerIDFile": "",
        "CpuShares": 4096,
        "CpusetCpus": "",
        "Devices": [],
        "Dns": null,
        "DnsSearch": null,
        "ExtraHosts": null,
        "IpcMode": "",
        "Links": null,
        "LogConfig": {
            "Config": null,
            "Type": "json-file"
        },
        "LxcConf": [],
        "Memory": 2147483648,
        "MemorySwap": 0,
        "NetworkMode": "bridge",
        "PidMode": "",
        "PortBindings": {},
        "Privileged": false,
        "PublishAllPorts": false,
        "ReadonlyRootfs": false,
        "RestartPolicy": {
            "MaximumRetryCount": 0,
            "Name": "no"
        },
        "SecurityOpt": null,
        "Ulimits": null,
        "VolumesFrom": null
    },
    "HostnamePath": "/var/lib/docker/containers/a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f/hostname",
    "HostsPath": "/var/lib/docker/containers/a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f/hosts",
    "Id": "a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f",
    "Image": "3554ec9321bfdd3ea85a270aa0009a0663e2a444f534f0c234ccb6d04d30a1ca",
    "LogPath": "/var/lib/docker/containers/a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f/a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f-json.log",
    "MountLabel": "",
    "Name": "/foo.bar",
    "NetworkSettings": {
        "Bridge": "docker0",
        "Gateway": "172.17.42.1",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "172.17.0.19",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "fe80::42:acff:fe11:13",
        "LinkLocalIPv6PrefixLen": 64,
        "MacAddress": "02:42:ac:11:00:13",
        "PortMapping": null,
        "Ports": {
            "10050/tcp": null,
            "7373/tcp": null,
            "7946/tcp": null,
            "9001/tcp": null
        }
    },
    "Path": "/usr/bin/supervisord",
    "ProcessLabel": "",
    "ResolvConfPath": "/var/lib/docker/containers/a8115e9632e799c1c4e3ccf5c935d085a857fa9d0bfdc899ed680eb08ba59f7f/resolv.conf",
    "RestartCount": 0,
    "State": {
        "Dead": false,
        "Error": "",
        "ExitCode": 0,
        "FinishedAt": "0001-01-01T00:00:00Z",
        "OOMKilled": false,
        "Paused": false,
        "Pid": 1652,
        "Restarting": false,
        "Running": true,
        "StartedAt": "2015-07-28T14:14:57.670936738Z"
    },
    "Volumes": {},
    "VolumesRW": {},
    "VolumesRelabel": {}
}
]
</code>

Any ideas? Do you need any more info?

**Comment by jangaraj** [comment]
Please enable debug mode of zabbix-agent(d):
```
DebugLevel=4
```
Restart zabbix-agent and provide start logs and logs during docker.discovery (docker.discovery[Name]). I think you have some problem with Additional Docker permissions.

**Comment by juwi** [comment]
Hi,

think I did find the reason.
First the requested outputs, though:

Discovery Log:
<code>
14568:20150730:143627.937 Requested [docker.discovery]
 14568:20150730:143627.937 In zbx_module_docker_discovery_basic()
 14568:20150730:143627.937 Sending back [{
        "data":[
                {
                        "{#FCONTAINERID}":"d49c292f3877767c74905ccbd265c9b3b83958b609d0cd24be8d66cc8ffa95f7",
                        "{#HCONTAINERID}":"d49c292f3877"},
                {
                        "{#FCONTAINERID}":"84a76f667210f632813b45e26f407a1854ef2681ee10c235c896b924d88ddfa0",
                        "{#HCONTAINERID}":"84a76f667210"},
                {
                        "{#FCONTAINERID}":"34e03e8a58efd185acf8b364cf060a0463494cfd82916818cef74cadd140db3a",
                        "{#HCONTAINERID}":"34e03e8a58ef"},
                {
                        "{#FCONTAINERID}":"1e141d3b75d2bc09c70df0938dda74b54ac560209220b8d1cbc2369e3425cd5d",
                        "{#HCONTAINERID}":"1e141d3b75d2"}]}]
</code>

Start Log:
<code>
25330:20150730:143917.361 Starting Zabbix Agent [Zabbix server]. Zabbix 2.2.7 (revision 50148).
 25330:20150730:143917.361 using configuration file: /etc/zabbix_agentd.conf
 25330:20150730:143917.361 In load_modules()
 25330:20150730:143917.361 loading module "/usr/lib/zabbix/modules/zabbix_module_docker.so"
 25330:20150730:143917.361 In zbx_module_api_version()
 25330:20150730:143917.361 In zbx_module_init()
 25330:20150730:143917.361 zabbix_module_docker v0.3.0, compilation time: Jul 28 2015 16:53:54
 25330:20150730:143917.361 In zbx_docker_dir_detect()
 25330:20150730:143917.361 Detected docker stat directory: /sys/fs/cgroup/
 25330:20150730:143917.362 Detected used docker driver dir: system.slice/
 25330:20150730:143917.362 Detected systemd docker - prefix/suffix will be used
 25330:20150730:143917.362 Detected JoinController cpu,cpuacct
 25330:20150730:143917.362 In zbx_docker_perm()
 25330:20150730:143917.362 Additional permission of Zabbix Agent are not detected - only basic docker metrics are availaible
</code>

I guess that last line is the reason. In zbx_docker_perm() the check should be
if((strcmp(gr->gr_name, "docker") == 0) || (strcmp(gr->gr_name, "dockerroot") == 0))
for this to work, correctly on RHEL/CentOS 7 based systems.

**Comment by jangaraj** [comment]
Thx for info. Could you provide more details about your docker (docker info), OS (release, kernel) and how did you install docker please?

I work with Centos/Rhel 7(.1) and docker 1.5 
and I didn't see this problem. Maybe it's customised docker or new docker release.

**Comment by juwi** [comment]
Hi,
it's CentOS 7.1 with Docker 1.6.2 from the CentOS Extras Repo. Contrary to the docker from dockerio's own repo it creates the group dockerroot instead of docker.

Here's docker info:
<code>
docker info
Containers: 23
Images: 475
Storage Driver: devicemapper
 Pool Name: docker-253:0-117179468-pool
 Pool Blocksize: 65.54 kB
 Backing Filesystem: extfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 14.59 GB
 Data Space Total: 107.4 GB
 Data Space Available: 92.78 GB
 Metadata Space Used: 23.83 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.124 GB
 Udev Sync Supported: true
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.93-RHEL7 (2015-01-28)
Execution Driver: native-0.2
Kernel Version: 3.10.0-229.7.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
CPUs: 12
Total Memory: 62.67 GiB
Name: foo
ID: bar
</code>

**Comment by jangaraj** [comment]
Root cause is problem with docker group (*dockerroot* is used, instead of *docker*) AND *docker.sock* permissions (only *root:root* is allowed to use *docker.sock*).

Solutions:
1.) Edit zabbix_agentd.conf and set AllowRoot (Zabbix agent with root permissions):
```
AllowRoot=1
```
2.) Use vanilla upstream Docker instead of RHEL/Centos Docker:
https://docs.docker.com/installation/centos/

3.) modify zabbix/docker user/group/permissions, so Zabbix will be member of standard docker group and it will be able to read/write docker.sock - I don't recommend it

For the record: see details from RHEL/Centor maintainer https://github.com/jangaraj/Zabbix-Docker-Monitoring/pull/18

Anyway thx for your issue @juwi :-). Hopefully, we have just saved some time of RHEL/Centos docker users in the future - README has been updated.
